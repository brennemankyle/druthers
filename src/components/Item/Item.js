import React from 'react'
import PropTypes from 'prop-types'
import AppPropTypes from '../../utils/AppPropTypes'

let Item = (props) => {
  return (
    <li className={props.className} val={props.item.value}>
      {props.item.label}
      {props.canRemove && <button className="remove" type="button" val={props.item.value} tabIndex="-1">
        <img className="remove" src="remove.png" alt="remove" val={props.item.value} /></button>}
    </li>
  )
}

Item.propTypes = {
  canRemove: PropTypes.bool.isRequired,
  item: AppPropTypes.item,
  multiple: PropTypes.bool.isRequired,
  styles: AppPropTypes.styles.isRequired,
}

export default Item
