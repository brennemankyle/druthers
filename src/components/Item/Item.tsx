import React, { ComponentType } from "react";
import { Item as ItemType } from "../../utils/SelectTypes";
import { RemoveProps } from "../svg/svg";

interface Props {
  className: string;
  removable: boolean;
  item: ItemType;
  svg_Remove: ComponentType<RemoveProps>;
}

function Item(props: Props) {
  let Remove = props.svg_Remove;

  return (
    <li className={props.className} data-val={props.item.value}>
      {props.item.displayElement ?? props.item.label}
      {props.removable && (
        <button
          className="remove"
          type="button"
          data-val={props.item.value}
          tabIndex={-1}
        >
          <Remove item={props.item} />
        </button>
      )}
    </li>
  );
}

export default Item;
