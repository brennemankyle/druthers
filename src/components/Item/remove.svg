<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" preserveAspectRatio="xMidYMid meet" viewBox="0 0 16 16">
  <defs>
    <path d="M10.87 8L16 13.12L13.13 16L8 10.87L2.87 16L0 13.12L5.13 8L0 2.87L2.87 0L8 5.12L13.13 0L16 2.87L10.87 8Z" id="aNsiuEacR"></path>
  </defs>
  <g>
    <g>
      <use xlink:href="#aNsiuEacR" opacity="1" fill="#000000" fill-opacity="1"></use>
    </g>
  </g>
</svg>
