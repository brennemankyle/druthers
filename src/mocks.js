let mockStyles = {
  colors: {}
}

let mockItem = {
  value: '42',
  label: 'Meaning of life',
}

let mockItemList = [
  {value: '1', label: 'Item 1'},
  {value: '2', label: 'Item 2'},
  {value: '3', label: 'Item 3'},
  {value: '4', label: 'Item 4'},
  {value: '5', label: 'Item 5'},
]

export { mockStyles, mockItem, mockItemList }
