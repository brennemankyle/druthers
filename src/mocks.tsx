import React from "react";
import defaultProps from "./utils/defaultProps";
import { withKeys, noop } from "./utils/utils";

let mockStyles = {
  styles_width: 600,
  styles_multiple: false,
  styles_disabled: false,
  styles_hasSelection: false,
  styles_hasOptions: true,
  styles_optionHighlighted: [],
  styles_rightToLeft: false,
  ...withKeys(defaultProps, "styles_"),
};

let mockItem = {
  value: "42",
  label: "Meaning of life",
};

let mockItemList = [
  { value: "1", label: "Item 1" },
  { value: "2", label: "Item 2" },
  { value: "3", label: "Item 3" },
  { value: "4", label: "Item 4" },
  { value: "5", label: "Item 5" },
];

let mockEvent = {
  preventDefault: noop,
};

let MockElement = (props: any) => <div data-props={props} />;
let MockInput = (props: any) => <input type="text" data-props={props} />;
let MockSvg = () => (
  <svg>
    <path d="M0 0L5 5Z" />
  </svg>
);

let mockRect = {
  top: 10,
  bottom: 110,
  left: 20,
  right: 220,
  width: 200,
  height: 100,
  x: 15,
  y: 25,
};

export {
  mockStyles,
  mockItem,
  mockItemList,
  mockEvent,
  MockElement,
  MockInput,
  MockSvg,
  mockRect,
};
