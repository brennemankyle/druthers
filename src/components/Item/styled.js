import styled from "@emotion/styled";
import Item from "./Item";

const Selection = styled(Item)`
  border-radius: ${props => props.styles_borderRadius};
  align-items: center;
  justify-content: space-between;
  margin-top: calc(${props => props.styles_selection_margin} / 2);
  margin-bottom: calc(${props => props.styles_selection_margin} / 2);
  padding-top: ${props => props.styles_selection_paddingTop};
  padding-bottom: ${props => props.styles_selection_paddingBottom};
  padding-left: ${props => props.styles_selection_paddingLeft};
  padding-right: ${props =>
    props.removable ? "0" : props.styles_selection_paddingRight};

  & .remove:not(svg):not(path) {
    display: flex;
    border: none;
    border-radius: ${props => props.styles_borderRadius};
    margin-top: -${props => props.styles_selection_paddingTop};
    margin-bottom: -${props => props.styles_selection_paddingBottom};
    padding-top: ${props => props.styles_selection_paddingTop};
    padding-bottom: ${props => props.styles_selection_paddingBottom};
    padding-left: ${props => props.styles_selection_paddingLeft};
    padding-right: ${props => props.styles_selection_paddingRight};
    background-color: transparent;
    font-size: 1em;

    svg {
      fill: ${props => props.styles_colors_primary};
      width: ${props => props.styles_icon_width};
      height: ${props => props.styles_icon_width};
    }

    ${props =>
      props.styles_selectionHighlighted === props.item.value &&
      `background-color: ${props.styles_colors_warning};

      svg {
        fill: ${props.styles_colors_warningBold};
      }`}
  }

  ${props =>
    props.styles_multiple
      ? `display: inline-flex;
      background-color: ${props.styles_colors_secondary};
      margin-right: ${props.styles_selection_margin};`
      : `display: flex;`}
  ${props =>
    props.styles_rightToLeft &&
    `
    transform: scaleX(-1);
    flex-direction: row-reverse;
    padding-left: ${
      props.removable ? "0" : props.styles_selection_paddingRight
    };
    padding-right: ${props.styles_selection_paddingLeft};`}
`;

const Option = styled(Item)`
  padding-top: ${props => props.styles_option_paddingTop};
  padding-bottom: ${props => props.styles_option_paddingBottom};
  padding-right: ${props => props.styles_option_paddingRight};
  padding-left: ${props => props.styles_option_paddingLeft};
  &:first-of-type{
    border-top-left-radius: ${props => props.styles_borderRadius};
    border-top-right-radius: ${props => props.styles_borderRadius};
  }
  &:last-of-type{
    border-bottom-left-radius: ${props => props.styles_borderRadius};
    border-bottom-right-radius: ${props => props.styles_borderRadius};
  }
  z-index: 999;

  ${props =>
    props.styles_optionHighlighted != null &&
    props.styles_optionHighlighted === props.item.value &&
    `background-color: ${props.styles_colors_highlight};`}
  ${props =>
    props.styles_rightToLeft &&
    `transform: scaleX(-1);
    text-align: right;`}
  ${props =>
    props.item.childGroup &&
    props.item.value == null &&
    `color: ${props.styles_colors_secondary};
    font-size: .8em;`}
  ${props =>
    props.item.childGroup && props.item.value != null && `font-weight: bold;`}
  ${props =>
    props.item.group &&
    `margin-left: ${props.item.group.split(".").length * 2}em;`}
`;

export { Selection, Option };
