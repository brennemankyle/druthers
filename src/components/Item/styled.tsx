import React from "react";
import styled from "@emotion/styled";
import Item from "./Item";

const Selection = styled(Item)(
  (props) => `
  color: ${props.styles_colors_text}
  border-radius: ${props.styles_borderRadius};
  align-items: center;
  justify-content: space-between;
  margin-top: calc(${props.styles_selection_margin} / 2);
  margin-bottom: calc(${props.styles_selection_margin} / 2);
  padding-top: ${props.styles_selection_paddingTop};
  padding-bottom: ${props.styles_selection_paddingBottom};
  padding-left: ${props.styles_selection_paddingLeft};
  padding-right: ${props.removable ? "0" : props.styles_selection_paddingRight};

  & .remove:not(svg):not(path) {
    display: flex;
    border: none;
    border-radius: ${props.styles_borderRadius};
    margin-top: -${props.styles_selection_paddingTop};
    margin-bottom: -${props.styles_selection_paddingBottom};
    padding-top: ${props.styles_selection_paddingTop};
    padding-bottom: ${props.styles_selection_paddingBottom};
    padding-left: ${props.styles_selection_paddingLeft};
    padding-right: ${props.styles_selection_paddingRight};
    background-color: transparent;
    font-size: 1em;

    svg {
      fill: ${props.styles_colors_primary};
      width: ${props.styles_icon_width};
      height: ${props.styles_icon_width};
    }

    ${
      props.styles_selectionHighlighted === props.item.value
        ? `background-color: ${props.styles_colors_warning};

      svg {
        fill: ${props.styles_colors_warningBold};
      }`
        : ``
    }
  }

  ${
    props.styles_multiple
      ? `display: inline-flex;
        background-color: ${props.styles_colors_secondary};
        margin-right: ${props.styles_selection_margin};`
      : `display: flex;`
  }
  ${
    props.styles_rightToLeft
      ? `
        transform: scaleX(-1);
        flex-direction: row-reverse;
        padding-left: ${
          props.removable ? "0" : props.styles_selection_paddingRight
        };
        padding-right: ${props.styles_selection_paddingLeft};`
      : ``
  }
`
);

const Option = styled(Item)(
  (props) => `
  color: ${props.styles_colors_text}
  padding-top: ${props.styles_option_paddingTop};
  padding-bottom: ${props.styles_option_paddingBottom};
  padding-right: ${props.styles_option_paddingRight};
  padding-left: ${props.styles_option_paddingLeft};
  &:first-of-type{
    border-top-left-radius: ${props.styles_borderRadius};
    border-top-right-radius: ${props.styles_borderRadius};
  }
  &:last-of-type{
    border-bottom-left-radius: ${props.styles_borderRadius};
    border-bottom-right-radius: ${props.styles_borderRadius};
  }
  z-index: 999;

  ${
    props.styles_optionHighlighted != null &&
    props.styles_optionHighlighted === props.item.value
      ? `background-color: ${props.styles_colors_highlight};`
      : ``
  }
  ${
    props.styles_rightToLeft
      ? `transform: scaleX(-1);
        text-align: right;`
      : ``
  }
  ${
    props.item.childGroup && props.item.value == null
      ? `color: ${props.styles_colors_secondary};
        font-size: .8em;`
      : ``
  }
  ${
    props.item.childGroup && props.item.value != null
      ? `font-weight: bold;`
      : ``
  }
  ${
    props.item.group
      ? `margin-left: ${props.item.group.split(".").length * 2}em;`
      : ``
  }
`
);

export { Selection, Option };
