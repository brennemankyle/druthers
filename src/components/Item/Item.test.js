import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { mockStyles, mockItem } from '../../mocks'
import Item from './Item'

it('renders', () => {
  const wrapper = shallow(<Item styles={mockStyles} item={mockItem} removable={false}></Item>)

  expect(toJson(wrapper)).toMatchSnapshot()
})

it('renders removable', () => {
  const wrapper = shallow(<Item styles={mockStyles} item={mockItem} removable={true}></Item>)

  expect(toJson(wrapper)).toMatchSnapshot()
})
