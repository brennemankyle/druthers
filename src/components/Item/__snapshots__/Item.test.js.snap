// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<li
  val="42"
>
  Meaning of life
</li>
`;

exports[`renders removable 1`] = `
<li
  val="42"
>
  Meaning of life
  <button
    className="remove"
    tabIndex="-1"
    type="button"
    val="42"
  >
    <img
      alt="remove"
      className="remove"
      src="remove.png"
      val="42"
    />
  </button>
</li>
`;
